File: python/src/ziyy/compiler/state.py
from ..value import *
class State:
    def __init__(self):
        self.tags = [""]
        self.saves = [RESET]

    def push(self, tag: str, string: str):
        l = len(self.saves) - 1
        s = self.saves[l]
        self.saves.append(s + string)
        self.tags.append(tag)

    def pop(self) -> tuple[str, str]:
        a = self.tags.pop()
        b = self.saves.pop()
        return a, b

    def current_tag(self) -> str:
        l = len(self.tags) - 1
        return self.tags[l]

    def current_save(self) -> str:
        l = len(self.saves) - 1
        return self.saves[l]

File: java/src/main/java/ziyy/compiler/State.java
package ziyy.compiler;

import java.util.ArrayList;

import static ziyy.value.Value.RESET;

public class State {
    ArrayList<String> tags = new ArrayList<>();
    ArrayList<String> saves = new ArrayList<>();

    public State() {
        tags.add("");
        saves.add(RESET);
    }

    public void push(String tag, String string) {
        int l = saves.size() - 1;
        String s = saves.get(l);
        saves.add(s + string);
        tags.add(tag);
    }

    public String[] pop() {
        int l = saves.size() - 1;
        String a = tags.remove(l);
        String b = saves.remove(l);

        String[] s = {a, b};
        return s;
    }

    public String currentTag() {
        int l = tags.size() - 1;
        return tags.get(l);
    }

    public String currentSave() {
        int l = saves.size() - 1;
        return saves.get(l);
    }
}
